File: python/kindergarten-garden/kindergarten_garden.py
"""Kindergarten Garden exercise."""

from typing import ClassVar


DEFAULT_STUDENTS = (
    "Alice",
    "Bob",
    "Charlie",
    "David",
    "Eve",
    "Fred",
    "Ginny",
    "Harriet",
    "Ileana",
    "Joseph",
    "Kincaid",
    "Larry",
)


class Garden:  # pylint: disable=too-few-public-methods
    """Representation for student's garden."""

    PLANTS: ClassVar = {
        "C": "Clover",
        "G": "Grass",
        "R": "Radishes",
        "V": "Violets",
    }

    PLANTS_PER_STUDENT_PER_ROW = 2
    PLANT_ROWS = 2, "two"

    def __init__(self, diagram, students=DEFAULT_STUDENTS):
        self._students = sorted(students)
        self._diagram = diagram.splitlines()
        n_rows, n_rows_str = self.PLANT_ROWS
        if (
            len(self._diagram) != n_rows
            or len({len(row) for row in self._diagram}) != 1
            or len(self._diagram[0]) % 2 != 0
        ):
            raise ValueError(
                f"Diagram should have {n_rows_str} even, equal length lines"
            )

    def plants(self, student):
        """List of student's plants."""
        pos = self._students.index(student) * self.PLANTS_PER_STUDENT_PER_ROW
        student_slice = slice(pos, pos + self.PLANTS_PER_STUDENT_PER_ROW)
        return [
            self.PLANTS[plant] for row in self._diagram for plant in row[student_slice]
        ]

File: c/kindergarten-garden/kindergarten_garden.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include "kindergarten_garden.h"

#define PLANT_AT(row, pos, offset) ((row)[2*(pos) + (offset)])
#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))
static plant_t read_plant(char plant);
static size_t get_student_position(const char *student);

plants_t plants(const char * diagram, const char *student) {
  size_t pos = get_student_position(student);
  const char *row1 = diagram;
  const char *first_newline = strpbrk(diagram, "\n");
  if (first_newline == NULL || *(first_newline+1) == '\0') {
    fputs("Second row not found!\n", stderr);
    exit(EXIT_FAILURE);
  }
  const char *row2 = first_newline + 1;
  return (plants_t){{
      read_plant(PLANT_AT(row1, pos, 0)),
      read_plant(PLANT_AT(row1, pos, 1)),
      read_plant(PLANT_AT(row2, pos, 0)),
      read_plant(PLANT_AT(row2, pos, 1))
    }};
}


static size_t get_student_position(const char *student) {
  static const char * students[] = { "Alice", "Bob", "Charlie", "David", "Eve",
                                     "Fred", "Ginny", "Harriet", "Ileana", "Joseph",
                                     "Kincaid", "Larry" };
  for(size_t k = 0; k < ARRAY_SIZE(students); k++) {
    if (0 == strncmp(students[k], student, strlen(students[k]))) {
      return k;
    }
  }
  return -1;
}

static plant_t read_plant(char plant) {
  switch(plant) {
  case 'C':
    return CLOVER;
  case 'G':
    return GRASS;
  case 'R':
    return RADISHES;
  case 'V':
    return VIOLETS;
  default:
    fprintf(stderr, "Unknown plant representation: '%c' (%d)\n", plant, (int) plant);
    exit(EXIT_FAILURE);
  }
}
